python-frontend/src/main/java/org/sonar/python/semantic/SymbolUtils.java

```diff
@@ -349,4 +349,40 @@ public static boolean isTypeShedFile(PythonFile pythonFile) {
     return pythonFile instanceof TypeShedPythonFile;
   }
 
+  /**
+   * @return the offset between parameter position and argument position:
+   *   0 if there is no implicit first parameter (self, cls, etc...)
+   *   1 if there is an implicit first parameter
+   *  -1 if unknown (intent is not clear from context)
+   */
+  public static int firstParameterOffset(FunctionSymbol functionSymbol, boolean isStaticCall) {
+    List<FunctionSymbol.Parameter> parameters = functionSymbol.parameters();
+    if (parameters.isEmpty()) {
+      return 0;
+    }
+    String firstParamName = parameters.get(0).name();
+    if (firstParamName == null) {
+      // First parameter is defined as a tuple
+      return -1;
+    }
+    List<String> decoratorNames = functionSymbol.decorators();
+    if (decoratorNames.size() > 1) {
+      // We want to avoid FP if there are many decorators
+      return -1;
+    }
+    if (!decoratorNames.isEmpty() && !decoratorNames.get(0).endsWith("classmethod") && !decoratorNames.get(0).endsWith("staticmethod")) {
+      // Unknown decorator which might alter the behaviour of the method
```

Comment:
It's true that this will most likely lead to a few FNs on the rules that use this method. That said, without knowing the behavior of the decorator, we really cannot be sure about this. 

It would be easy to create a decorator that passes additional arguments/remove some in some cases, so I think it's a safe approach to avoid concluding in this case (otherwise that could lead to FPs in the related rules). Maybe later we can try to gather the fully qualified names of decorators that we know don't change the parameter behaviors, to have something more robust.